File: Leetcode/1498. Number of Subsequences That Satisfy the Given Sum Condition/code.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    int numSubseq(vector<int>& nums, int target) {
        const int MOD = 1e9 + 7; 
        sort(nums.begin(), nums.end()); 
        int n = nums.size();
        vector<int> pow2(n, 1); 
        for (int i = 1; i < n; ++i) {
            pow2[i] = (pow2[i - 1] * 2) % MOD;
        }

        int left = 0, right = n - 1;
        int count = 0;

        while (left <= right) {
            if (nums[left] + nums[right] <= target) {
                count = (count + pow2[right - left]) % MOD; 
                left++; 
            } else {
                right--; 
            }
        }

        return count;
    }
};

int main() {
    Solution sol;
    vector<int> nums = {3, 5, 6, 7};
    int target = 9;

    cout << "Number of valid subsequences: " << sol.numSubseq(nums, target) << endl;
    return 0;
}

File: Leetcode/1498. Number of Subsequences That Satisfy the Given Sum Condition/code.py
class Solution:
    def numSubseq(self, nums, target):
        MOD = 10**9 + 7
        nums.sort()  
        n = len(nums)

        pow2 = [1] * n
        for i in range(1, n):
            pow2[i] = (pow2[i - 1] * 2) % MOD

        left, right = 0, n - 1
        count = 0

        while left <= right:
            if nums[left] + nums[right] <= target:
                count = (count + pow2[right - left]) % MOD  
                left += 1  
            else:
                right -= 1  

        return count


# Example usage
sol = Solution()
nums = [3, 5, 6, 7]
target = 9
print("Number of valid subsequences:", sol.numSubseq(nums, target))
